browser-test/src/support/index.ts

```diff
@@ -82,6 +89,96 @@ export const startSession = async (): Promise<{
   return {browser, context, page}
 }
 
+/**
+ * Object containing properties and methods for interacting with browser and
+ * app. See docs for createBrowserContext() method for more info.
+ */
+export interface TestContext {
+  /**
+   * Playwright Page object. Provides functionality to directly interact with
+   * the browser .
+   * Methods: https://playwright.dev/docs/api/class-page
+   */
+  page: Page
+
+  adminQuestions: AdminQuestions
+  adminPrograms: AdminPrograms
+}
+
+/**
+ * Launches a browser and returns context that contains objects needed to
+ * interact with the browser. Example usage:
+ *
+ * describe('some test', () => {
+ *   const ctx = createBrowserContext()
+ *
+ *   it('should do foo', async () => {
+ *     await ctx.page.click('#some-button')
+ *   })
+ * })
+ *
+ * Browser session is reset between tests and database is cleared by default.
+ * Each test starts on the login page.
+ *
+ * Context object should be accessed only from within it(), before/afterEach(),
+ * before/afterAll() functions.
+ *
+ * @param clearDb Whether database is cleared between tests. True by default.
+ *     It's recommended that database is cleared between tests to keep tests
+ *     hermetic.
+ * @return object containing browser page. Context object is reset between tests
+ *     so none of its properties should be cached and reused between tests.
+ */
+export const createTestContext = (clearDb = true): TestContext => {
+  let browser: Browser
+  let browserContext: BrowserContext
+
+  // TestContext properties are set in resetContext() later. For now we just
+  // need an object that we can return to caller. Caller is expected to access
+  // it only from before/afterX functions or tests.
+  const ctx: TestContext = {} as unknown as TestContext
+
+  // We create new browser context and session before each test. It's
+  // important to get fresh browser context so that each test gets its own
+  // videos. If we reuse same browser context - we'll get one huge video for
+  // all tests.
+  async function resetContext() {
+    if (browserContext != null) {
+      await browserContext.close()
+    }
+    browserContext = await makeBrowserContext(browser)
+    ctx.page = await browserContext.newPage()
+    ctx.adminQuestions = new AdminQuestions(ctx.page)
+    ctx.adminPrograms = new AdminPrograms(ctx.page)
+    await ctx.page.goto(BASE_URL)
+  }
+
+  beforeAll(async () => {
+    browser = await chromium.launch()
+    await resetContext()
+  })
+
+  beforeEach(async () => {
+    await resetContext()
+  })
+
+  afterEach(async () => {
+    if (clearDb) {
+      await dropTables(ctx.page)
```

Comment:
Some of our tests have nested describes, e.g.
```
describe('Applicant flow', () => {
  describe('Checkbox q', () => {
    it ('test case 1', () => {...})
    it ('test case 2', () => {...})
  })

  describe('Address q', () => {
    it ('test case 1', () => {...})
    it ('test case 2', () => {...})
  })
})
```

In this case, is the intention that the `const ctx = createTestContext()` goes in the innermost describe?

Reply:
No, `const ctx = createTestContext()` goes at the very beginning of the top-most describe. Added comment in the jsdoc about it.

Reply:
Ah I'm sorry, I was misunderstanding how beforeEach worked. I thought it would only be on the outer level describe, but I see now that it will run before each nested "it". Ok this makes sense then, so we drop the tables after each test.